src/ucp/tag/offload.c

```diff
@@ -144,61 +191,40 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_tag_offload_unexp_rndv,
 {
     ucp_worker_t *worker        = arg;
     ucp_request_hdr_t *rndv_hdr = (ucp_request_hdr_t*)hdr;
-    ucp_ep_t *ep                = ucp_worker_get_reply_ep(worker, rndv_hdr->sender_uuid);
-    size_t len                  = sizeof(ucp_rndv_rts_hdr_t);
     ucp_rndv_rts_hdr_t *rts;
-    ucp_sw_rndv_hdr_t  *sw_rndv_hdr;
-    const uct_md_attr_t *md_attr;
+    size_t len;
     size_t rkey_size;
 
-    if (!ucp_ep_is_stub(ep) && rkey_buf) {
-        md_attr   = ucp_ep_md_attr(ep, ucp_ep_get_tag_lane(ep));
-        rkey_size = md_attr->rkey_packed_size;
-        len += rkey_size;
-    } else {
-        rkey_size = 0;
-    }
-
-    rts = ucs_alloca(len);
-
-
-    /* Fill RTS to emulate SW RNDV flow. */
-    rts->super.tag = stag;
-    rts->sreq      = *rndv_hdr;
-    rts->address   = remote_addr;
-
+    rkey_size = ucp_tag_offload_rkey_size(worker->context);
+    rts       = ucs_alloca(sizeof(*rts) + rkey_size); /* SW rndv req may also
+                                                         carry a key */
     if (remote_addr) {
-        rts->size  = length;
-        rts->flags = UCP_RNDV_RTS_FLAG_OFFLOAD;
-        if (rkey_size) {
-            memcpy(rts + 1, rkey_buf, rkey_size);
-            rts->flags |= UCP_RNDV_RTS_FLAG_PACKED_RKEY;
-        }
+        /* Unexpected tag offload RNDV */
+        ucp_tag_offload_fill_rts(rts, rndv_hdr, stag, remote_addr, length, 0);
+        len = ucp_tag_offload_copy_rkey(rts, (void*)rkey_buf, rkey_size) +
+              sizeof(*rts);
+
         UCP_WORKER_STAT_TAG_OFFLOAD(worker, RX_UNEXP_RNDV);
     } else {
-        /* This must be SW RNDV request. Take length from its header. */
-        sw_rndv_hdr = ucs_derived_of(hdr, ucp_sw_rndv_hdr_t);
-        rts->size   = sw_rndv_hdr->length;
-        rts->flags  = 0;
-        len        += ucp_tag_offload_fetch_rkey(ep, sw_rndv_hdr, rts);
+        /* Unexpected tag offload rndv request. Sender buffer is either
+           non-contig or it's length > rndv.max_zcopy capability of tag lane */
+        len = ucp_tag_offload_fill_sw_rts(rts, (void*)hdr, hdr_length, stag,
```

Comment:
seems inconsistent, this function gets rkey_size, which should be 0 in this case (since remote_addr==0), but inside the function it checks UCP_RNDV_RTS_FLAG_PACKED_RKEY bit

Reply:
No, rkey_size is taken from md attrs in the very beginning (`rkey_size = ucp_tag_offload_rkey_size(worker->context`).
`UCP_RNDV_RTS_FLAG_PACKED_RKEY` bit maybe present in SW RNDV hdr as well, when RNDV messages is contiguous, but does not fit a single network operation. In this case we pass an address and rkey for remote peer to make RNDV in SW with RNDV_GETs